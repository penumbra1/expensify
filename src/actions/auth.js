import { firebase, googleAuthProvider } from "../firebase/firebase";

export const login = uid => ({ type: "LOGIN", uid });

export const startLogin = () => dispatch =>
  firebase.auth().signInWithPopup(googleAuthProvider);

export const logout = () => ({ type: "LOGOUT" });

export const startLogout = () => dispatch => firebase.auth().signOut();
