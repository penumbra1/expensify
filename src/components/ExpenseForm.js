import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const newDescription = e.target.value;
    this.setState(() => ({ description: newDescription }));
  };

  onNoteChange = e => {
    const newNote = e.target.value;
    this.setState(() => ({ note: newNote }));
  };

  onAmountChange = e => {
    const newAmount = e.target.value;
    if (!newAmount || newAmount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount: newAmount }));
    }
  };

  onDateChange = newDate => {
    if (newDate) {
      this.setState(() => ({ createdAt: newDate }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onFormSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // Error message
      this.setState(() => ({
        error: "Please provide a description and an amount"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onFormSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            id="datePicker"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>
            {this.props.expense ? "Update expense" : "Add expense"}
          </button>
        </form>
      </div>
    );
  }
}
