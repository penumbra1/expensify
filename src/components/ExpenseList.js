import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = ({ expenses }) =>
  expenses.length === 0 ? (
    <p>No expenses</p>
  ) : (
    expenses.map(expense => <ExpenseListItem {...expense} key={expense.id} />)
  );

export const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
