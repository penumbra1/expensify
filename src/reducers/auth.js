export default (state = {}, action) => {
  switch (action.type) {
    case "LOGIN":
      return { uid: action.uid };
    case "LOGOUT":
      return { uid: null };
    default:
      return state;
  }
};
