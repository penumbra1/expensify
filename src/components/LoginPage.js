import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

export const LoginPage = props => (
  <button onClick={props.startLogin}>Log in</button>
);

const mapDispatchToProps = { startLogin };

export default connect(
  undefined,
  mapDispatchToProps
)(LoginPage);
