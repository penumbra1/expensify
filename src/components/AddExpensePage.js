import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import { DASHBOARD } from "../routers/pathNames";

// Export the unconnected component as well for testing
// independently from Redux and dispatch (with spies instead of prop methods)
export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push(DASHBOARD);
  };

  render() {
    return (
      <Fragment>
        <h1>Add an expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </Fragment>
    );
  }
}

// Map dispatch to an addExpense prop
// to abstract it away from the component
const mapDispatchToProps = { startAddExpense };

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
