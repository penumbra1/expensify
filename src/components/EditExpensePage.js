import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import NotFoundPage from "./NotFoundPage";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import { DASHBOARD } from "../routers/pathNames";

// Export the unconnected component as well for testing
// independently from Redux and dispatch (with spies instead of prop methods)
export class EditExpensePage extends Component {
  onSubmit = updates => {
    this.props.startEditExpense(this.props.expense.id, updates);
    this.props.history.push(DASHBOARD);
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push(DASHBOARD);
  };

  render() {
    // Redirect to 404 page if no expense was found (URL or id incorrect)
    if (!this.props.expense) {
      return <NotFoundPage />;
    }
    return (
      <Fragment>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </Fragment>
    );
  }
}

export const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = {
  startEditExpense,
  startRemoveExpense
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
