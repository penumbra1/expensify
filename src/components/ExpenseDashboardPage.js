import React, { Fragment } from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboardPage = () => (
  <Fragment>
    <ExpenseListFilters />
    <ExpenseList />
  </Fragment>
);

export default ExpenseDashboardPage;
