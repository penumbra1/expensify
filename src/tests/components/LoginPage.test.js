import React from "react";
import { shallow } from "enzyme";
import { LoginPage } from "../../components/LoginPage";
import { startLogin } from "../../firebase/auth";

jest.mock("../../firebase/auth", () => ({ startLogin: jest.fn() }));

test("should render login page correctly", () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});

test("should call startLogin on button click", () => {
  const wrapper = shallow(<LoginPage />);
  wrapper.find("button").simulate("click");

  expect(startLogin).toHaveBeenCalled();
});
