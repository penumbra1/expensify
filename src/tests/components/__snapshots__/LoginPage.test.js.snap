// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login page correctly 1`] = `
<button
  onClick={[Function]}
>
  Log in
</button>
`;
